// @ybd-project/ytdl-core@6.0.1/package/YtdlCore.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.YtdlCore = void 0;
const ua_parser_js_1 = __importDefault(require("ua-parser-js"));
const Platform_1 = require("./platforms/Platform");
const Download_1 = require("./core/Download");
const Info_1 = require("./core/Info");
const Html5Player_1 = require("./core/Info/parser/Html5Player");
const OAuth2_1 = require("./core/OAuth2");
const errors_1 = require("./core/errors");
const Url_1 = require("./utils/Url");
const Format_1 = require("./utils/Format");
const Constants_1 = require("./utils/Constants");
const Log_1 = require("./utils/Log");
const SHIM = Platform_1.Platform.getShim(), Cache = SHIM.cache, FileCache = SHIM.fileCache;
function isNodeVersionOk(version) {
    return parseInt(version.replace('v', '').split('.')[0]) >= 16;
}
function isBrowserVersionOk(userAgent) {
    const PARSED = new ua_parser_js_1.default(userAgent).getBrowser(), VERSION = PARSED.version || '', BROWSER_NAME = PARSED.name?.toLocaleLowerCase() || '';
    if (BROWSER_NAME.includes('chrome')) {
        return VERSION.startsWith('76');
    }
    else if (BROWSER_NAME.includes('edge')) {
        return VERSION.startsWith('80');
    }
    else if (BROWSER_NAME.includes('firefox')) {
        return VERSION.startsWith('78');
    }
    else if (BROWSER_NAME.includes('safari')) {
        return VERSION.startsWith('14');
    }
    else if (BROWSER_NAME.includes('brave')) {
        return VERSION.startsWith('1');
    }
    else if (BROWSER_NAME.includes('opera')) {
        return VERSION.startsWith('63');
    }
    else {
        Log_1.Logger.debug('[ BrowserCheck ]: The specified UserAgent did not match the following browsers. (Chrome, Edge, FireFox, Safari, Brave, Opera)');
    }
    return true;
}
class YtdlCore {
    /* Constructor */
    constructor({ hl, gl, rewriteRequest, poToken, disablePoTokenAutoGeneration, visitorData, includesPlayerAPIResponse, includesNextAPIResponse, includesOriginalFormatData, includesRelatedVideo, clients, disableDefaultClients, oauth2Credentials, parsesHLSFormat, originalProxy, quality, filter, excludingClients, includingClients, range, begin, liveBuffer, highWaterMark, IPv6Block, dlChunkSize, streamType, disableBasicCache, disableFileCache, fetcher, logDisplay, noUpdate } = {}) {
        /* Get Info Options */
        this.hl = 'en';
        this.gl = 'US';
        this.disablePoTokenAutoGeneration = false;
        this.includesPlayerAPIResponse = false;
        this.includesNextAPIResponse = false;
        this.includesOriginalFormatData = false;
        this.includesRelatedVideo = true;
        this.disableDefaultClients = false;
        this.oauth2 = null;
        this.parsesHLSFormat = false;
        this.excludingClients = [];
        this.includingClients = 'all';
        this.streamType = 'default';
        /* Metadata */
        this.version = Constants_1.VERSION;
        const SHIM = Platform_1.Platform.getShim();
        /* Other Options */
        const LOG_DISPLAY = (logDisplay === 'none' ? [] : logDisplay) || ['info', 'success', 'warning', 'error'];
        SHIM.options.other.logDisplay = LOG_DISPLAY;
        Log_1.Logger.logDisplay = LOG_DISPLAY;
        SHIM.options.other.noUpdate = noUpdate ?? false;
        if (fetcher) {
            SHIM.fetcher = fetcher;
            SHIM.requestRelated.originalProxy = originalProxy;
            SHIM.requestRelated.rewriteRequest = rewriteRequest;
        }
        if (disableBasicCache) {
            Cache.disable();
        }
        if (disableFileCache) {
            FileCache.disable();
        }
        /* Get Info Options */
        this.hl = hl || 'en';
        this.gl = gl || 'US';
        this.rewriteRequest = rewriteRequest || undefined;
        this.disablePoTokenAutoGeneration = disablePoTokenAutoGeneration ?? false;
        this.includesPlayerAPIResponse = includesPlayerAPIResponse ?? false;
        this.includesNextAPIResponse = includesNextAPIResponse ?? false;
        this.includesOriginalFormatData = includesOriginalFormatData ?? false;
        this.includesRelatedVideo = includesRelatedVideo ?? true;
        this.clients = clients || undefined;
        this.disableDefaultClients = disableDefaultClients ?? false;
        this.parsesHLSFormat = parsesHLSFormat ?? false;
        this.originalProxy = originalProxy || undefined;
        if (this.originalProxy) {
            Log_1.Logger.debug(`<debug>"${this.originalProxy.base}"</debug> is used for <blue>API requests</blue>.`);
            Log_1.Logger.debug(`<debug>"${this.originalProxy.download}"</debug> is used for <blue>video downloads</blue>.`);
            Log_1.Logger.debug(`The query name <debug>"${this.originalProxy.urlQueryName || 'url'}"</debug> is used to specify the URL in the request. <blue>(?url=...)</blue>`);
        }
        this.setPoToken(poToken);
        this.setVisitorData(visitorData);
        this.setOAuth2(oauth2Credentials || null);
        /* Format Selection Options */
        this.quality = quality || undefined;
        this.filter = filter || undefined;
        this.excludingClients = excludingClients || [];
        this.includingClients = includingClients || 'all';
        /* Download Options */
        this.range = range || undefined;
        this.begin = begin || undefined;
        this.liveBuffer = liveBuffer || undefined;
        this.highWaterMark = highWaterMark || undefined;
        this.IPv6Block = IPv6Block || undefined;
        this.dlChunkSize = dlChunkSize || undefined;
        this.streamType = streamType || 'default';
        if (!this.disablePoTokenAutoGeneration) {
            this.automaticallyGeneratePoToken();
        }
        this.initializeHtml5PlayerCache();
        /* Version Check */
        if (SHIM.runtime === 'default') {
            if (!isNodeVersionOk(process.version)) {
                throw new errors_1.PlatformError(`You are using Node.js ${process.version} which is not supported. Minimum version required is v16.`);
            }
        }
        else if (SHIM.runtime === 'browser') {
            if (!isBrowserVersionOk(navigator.userAgent)) {
                throw new errors_1.PlatformError(`The environment in which ytdl-core is currently running is not supported. ytdl-core has a minimum version that works with each browser due to limitations of the API it uses internally. Please see the README for details.`);
            }
        }
        /* Load */
        SHIM.options.download = { hl: this.hl, gl: this.gl, rewriteRequest: this.rewriteRequest, poToken: this.poToken, disablePoTokenAutoGeneration: this.disablePoTokenAutoGeneration, visitorData: this.visitorData, includesPlayerAPIResponse: this.includesPlayerAPIResponse, includesNextAPIResponse: this.includesNextAPIResponse, includesOriginalFormatData: this.includesOriginalFormatData, includesRelatedVideo: this.includesRelatedVideo, clients: this.clients, disableDefaultClients: this.disableDefaultClients, oauth2Credentials, parsesHLSFormat: this.parsesHLSFormat, originalProxy: this.originalProxy, quality: this.quality, filter: this.filter, excludingClients: this.excludingClients, includingClients: this.includingClients, range: this.range, begin: this.begin, liveBuffer: this.liveBuffer, highWaterMark: this.highWaterMark, IPv6Block: this.IPv6Block, dlChunkSize: this.dlChunkSize };
        Platform_1.Platform.load(SHIM);
    }
    /* Setup */
    async setPoToken(poToken) {
        const PO_TOKEN_CACHE = await FileCache.get('poToken');
        if (poToken) {
            this.poToken = poToken;
        }
        else if (PO_TOKEN_CACHE) {
            Log_1.Logger.debug('PoToken loaded from cache.');
            this.poToken = PO_TOKEN_CACHE || undefined;
        }
        FileCache.set('poToken', this.poToken || '', { ttl: 60 * 60 * 24 });
    }
    async setVisitorData(visitorData) {
        const VISITOR_DATA_CACHE = await FileCache.get('visitorData');
        if (visitorData) {
            this.visitorData = visitorData;
        }
        else if (VISITOR_DATA_CACHE) {
            Log_1.Logger.debug('VisitorData loaded from cache.');
            this.visitorData = VISITOR_DATA_CACHE || undefined;
        }
        FileCache.set('visitorData', this.visitorData || '', { ttl: 60 * 60 * 24 });
    }
    async setOAuth2(oauth2Credentials) {
        const OAUTH2_CACHE = (await FileCache.get('oauth2')) || undefined;
        try {
            if (oauth2Credentials) {
                this.oauth2 = new OAuth2_1.OAuth2(oauth2Credentials) || undefined;
            }
            else if (OAUTH2_CACHE) {
                this.oauth2 = new OAuth2_1.OAuth2(OAUTH2_CACHE);
            }
            else {
                this.oauth2 = null;
            }
        }
        catch {
            this.oauth2 = null;
        }
    }
    automaticallyGeneratePoToken() {
        if (!this.poToken && !this.visitorData) {
            Log_1.Logger.debug('Since PoToken and VisitorData are <warning>not specified</warning>, they are generated <info>automatically</info>.');
            const generatePoToken = Platform_1.Platform.getShim().poToken;
            generatePoToken()
                .then(({ poToken, visitorData }) => {
                this.poToken = poToken;
                this.visitorData = visitorData;
                FileCache.set('poToken', this.poToken || '', { ttl: 60 * 60 * 24 });
                FileCache.set('visitorData', this.visitorData || '', { ttl: 60 * 60 * 24 });
            })
                .catch(() => { });
        }
    }
    initializeHtml5PlayerCache() {
        const HTML5_PLAYER = FileCache.get('html5Player');
        if (!HTML5_PLAYER) {
            Log_1.Logger.debug('To speed up processing, html5Player and signatureTimestamp are pre-fetched and cached.');
            (0, Html5Player_1.getHtml5Player)({});
        }
    }
    initializeOptions(options) {
        const INTERNAL_OPTIONS = { ...options, oauth2: this.oauth2 };
        INTERNAL_OPTIONS.hl = options.hl || this.hl;
        INTERNAL_OPTIONS.gl = options.gl || this.gl;
        INTERNAL_OPTIONS.rewriteRequest = options.rewriteRequest || this.rewriteRequest;
        INTERNAL_OPTIONS.poToken = options.poToken || this.poToken;
        INTERNAL_OPTIONS.disablePoTokenAutoGeneration = options.disablePoTokenAutoGeneration || this.disablePoTokenAutoGeneration;
        INTERNAL_OPTIONS.visitorData = options.visitorData || this.visitorData;
        INTERNAL_OPTIONS.includesPlayerAPIResponse = options.includesPlayerAPIResponse || this.includesPlayerAPIResponse;
        INTERNAL_OPTIONS.includesNextAPIResponse = options.includesNextAPIResponse || this.includesNextAPIResponse;
        INTERNAL_OPTIONS.includesOriginalFormatData = options.includesOriginalFormatData || this.includesOriginalFormatData;
        INTERNAL_OPTIONS.includesRelatedVideo = options.includesRelatedVideo || this.includesRelatedVideo;
        INTERNAL_OPTIONS.clients = options.clients || this.clients;
        INTERNAL_OPTIONS.disableDefaultClients = options.disableDefaultClients || this.disableDefaultClients;
        INTERNAL_OPTIONS.oauth2Credentials = options.oauth2Credentials || this.oauth2?.getCredentials();
        INTERNAL_OPTIONS.parsesHLSFormat = options.parsesHLSFormat || this.parsesHLSFormat;
        INTERNAL_OPTIONS.originalProxy = options.originalProxy || this.originalProxy || undefined;
        /* Format Selection Options */
        INTERNAL_OPTIONS.quality = options.quality || this.quality || undefined;
        INTERNAL_OPTIONS.filter = options.filter || this.filter || undefined;
        INTERNAL_OPTIONS.excludingClients = options.excludingClients || this.excludingClients || [];
        INTERNAL_OPTIONS.includingClients = options.includingClients || this.includingClients || 'all';
        /* Download Options */
        INTERNAL_OPTIONS.range = options.range || this.range || undefined;
        INTERNAL_OPTIONS.begin = options.begin || this.begin || undefined;
        INTERNAL_OPTIONS.liveBuffer = options.liveBuffer || this.liveBuffer || undefined;
        INTERNAL_OPTIONS.highWaterMark = options.highWaterMark || this.highWaterMark || undefined;
        INTERNAL_OPTIONS.IPv6Block = options.IPv6Block || this.IPv6Block || undefined;
        INTERNAL_OPTIONS.dlChunkSize = options.dlChunkSize || this.dlChunkSize || undefined;
        INTERNAL_OPTIONS.streamType = options.streamType || this.streamType || 'default';
        if (!INTERNAL_OPTIONS.oauth2 && options.oauth2Credentials) {
            Log_1.Logger.warning('The OAuth2 token should be specified when instantiating the YtdlCore class, not as a function argument.');
            INTERNAL_OPTIONS.oauth2 = new OAuth2_1.OAuth2(options.oauth2Credentials);
        }
        return INTERNAL_OPTIONS;
    }
    /** TIP: The options specified in new YtdlCore() are applied by default. (The function arguments specified will take precedence.) */
    download(link, options = {}) {
        return (0, Download_1.download)(link, this.initializeOptions(options));
    }
    /** TIP: The options specified in new YtdlCore() are applied by default. (The function arguments specified will take precedence.) */
    downloadFromInfo(info, options = {}) {
        return (0, Download_1.downloadFromInfo)(info, this.initializeOptions(options));
    }
    /** TIP: The options specified in new YtdlCore() are applied by default. (The function arguments specified will take precedence.) */
    getBasicInfo(link, options = {}) {
        return (0, Info_1.getBasicInfo)(link, this.initializeOptions(options));
    }
    /** TIP: The options specified in new YtdlCore() are applied by default. (The function arguments specified will take precedence.) */
    getFullInfo(link, options = {}) {
        return (0, Info_1.getFullInfo)(link, this.initializeOptions(options));
    }
}
exports.YtdlCore = YtdlCore;
YtdlCore.chooseFormat = Format_1.FormatUtils.chooseFormat;
YtdlCore.filterFormats = Format_1.FormatUtils.filterFormats;
YtdlCore.validateID = Url_1.Url.validateID;
YtdlCore.validateURL = Url_1.Url.validateURL;
YtdlCore.getURLVideoID = Url_1.Url.getURLVideoID;
YtdlCore.getVideoID = Url_1.Url.getVideoID;
